core/src/trezor/ui/layouts/common.py

```diff
@@ -1,42 +1,98 @@
 from typing import TYPE_CHECKING
 
-from trezor import workflow
+import trezorui2
+from trezor import ui, utils, workflow
 from trezor.enums import ButtonRequestType
 from trezor.messages import ButtonAck, ButtonRequest
-from trezor.wire import context
+from trezor.wire import ActionCancelled, context
 
 if TYPE_CHECKING:
-    from typing import Awaitable, Protocol, TypeVar
+    from typing import Any, Awaitable, Callable, TypeVar
 
-    T = TypeVar("T")
-
-    LayoutType = Awaitable
     PropertyType = tuple[str | None, str | bytes | None]
     ExceptionType = BaseException | type[BaseException]
 
-    class ProgressLayout(Protocol):
-        def report(self, value: int, description: str | None = None) -> None: ...
+    InfoFunc = Callable[[], Awaitable[None]]
 
+    T = TypeVar("T")
 
-async def button_request(
+
+async def _button_request(
     br_name: str,
     code: ButtonRequestType = ButtonRequestType.Other,
-    pages: int | None = None,
+    pages: int = 0,
 ) -> None:
     workflow.close_others()
     await context.maybe_call(
-        ButtonRequest(code=code, pages=pages, name=br_name), ButtonAck
+        ButtonRequest(code=code, pages=pages or None, name=br_name), ButtonAck
     )
 
 
 async def interact(
-    layout: LayoutType[T],
-    br_name: str,
+    layout_obj: ui.LayoutObj[T],
+    br_name: str | None,
     br_code: ButtonRequestType = ButtonRequestType.Other,
+    raise_on_cancel: ExceptionType | None = ActionCancelled,
 ) -> T:
-    pages = None
-    if hasattr(layout, "page_count") and layout.page_count() > 1:  # type: ignore [Cannot access attribute "page_count" for class "LayoutType"]
-        # We know for certain how many pages the layout will have
-        pages = layout.page_count()  # type: ignore [Cannot access attribute "page_count" for class "LayoutType"]
-    await button_request(br_name, br_code, pages)
-    return await layout
+    # shut down other workflows to prevent them from interfering with the current one
+    workflow.close_others()
+    # start the layout
+    layout = ui.Layout(layout_obj)
+    layout.start()
+    # send the button request
+    if br_name is not None:
+        await _button_request(br_name, br_code, layout_obj.page_count())
+    # wait for the layout result
+    result = await layout.get_result()
+    # raise an exception if the user cancelled the action
+    if raise_on_cancel is not None and result is trezorui2.CANCELLED:
+        raise raise_on_cancel
+    return result
+
+
+def raise_if_not_confirmed(
+    layout_obj: ui.LayoutObj[ui.UiResult],
+    br_name: str | None,
+    br_code: ButtonRequestType = ButtonRequestType.Other,
+    exc: ExceptionType = ActionCancelled,
+) -> Awaitable[None]:
+    action = interact(layout_obj, br_name, br_code, exc)
+    return action  # type: ignore ["UiResult" is incompatible with "None"]
```

Comment:
we're not returning `UiResult` though, we are returning an `Awaitable`.
the semantics of function "raise if not confirmed" are "don't care about the result value" so we need to mask the fact that the returned Awaitable is technically returning a value.
(the other way would be to propagate the `UiResult` through sometimes rather deep call stacks -- which is ultimately pointless because those stacks don't actually care about the result value)